exchanges/kucoin/kucoin_wrapper.go

```diff
@@ -0,0 +1,1443 @@
+				return err
+			}
+		}
+		return nil
+	case asset.Spot, asset.Margin:
+		ticks, err := ku.GetTickers(ctx)
+		if err != nil {
+			return err
+		}
+		pairs, err := ku.GetEnabledPairs(assetType)
+		if err != nil {
+			return err
+		}
+		for t := range ticks.Tickers {
+			pair, err := currency.NewPairFromString(ticks.Tickers[t].Symbol)
+			if err != nil {
+				return err
+			}
+			if !pairs.Contains(pair, true) {
+				continue
+			}
+			tick := ticker.Price{
+				Last:         ticks.Tickers[t].Last,
+				High:         ticks.Tickers[t].High,
+				Low:          ticks.Tickers[t].Low,
+				Volume:       ticks.Tickers[t].Volume,
+				Ask:          ticks.Tickers[t].Sell,
+				Bid:          ticks.Tickers[t].Buy,
+				Pair:         pair,
+				ExchangeName: ku.Name,
+				AssetType:    assetType,
+				LastUpdated:  ticks.Time.Time(),
+			}
+			assetEnabledPairs := ku.listOfAssetsCurrencyPairEnabledFor(pair)
+			if assetEnabledPairs[asset.Spot] && ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
+				err = ticker.ProcessTicker(&tick)
+				if err != nil {
+					return err
+				}
+			}
+			if assetEnabledPairs[asset.Margin] && ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
+				marginTick := tick
+				marginTick.AssetType = asset.Margin
+				err = ticker.ProcessTicker(&marginTick)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	default:
+		return fmt.Errorf("%w %v", asset.ErrNotSupported, assetType)
+	}
+	return nil
+}
+
+// FetchTicker returns the ticker for a currency pair
+func (ku *Kucoin) FetchTicker(ctx context.Context, p currency.Pair, assetType asset.Item) (*ticker.Price, error) {
+	p, err := ku.FormatExchangeCurrency(p, assetType)
+	if err != nil {
+		return nil, err
+	}
+	tickerNew, err := ticker.GetTicker(ku.Name, p, assetType)
+	if err != nil {
+		return ku.UpdateTicker(ctx, p, assetType)
+	}
+	return tickerNew, nil
+}
+
+// FetchOrderbook returns orderbook base on the currency pair
+func (ku *Kucoin) FetchOrderbook(ctx context.Context, pair currency.Pair, assetType asset.Item) (*orderbook.Base, error) {
+	pair, err := ku.FormatExchangeCurrency(pair, assetType)
+	if err != nil {
+		return nil, err
+	}
+	ob, err := orderbook.Get(ku.Name, pair, assetType)
+	if err != nil {
+		return ku.UpdateOrderbook(ctx, pair, assetType)
+	}
+	return ob, nil
+}
+
+// UpdateOrderbook updates and returns the orderbook for a currency pair
+func (ku *Kucoin) UpdateOrderbook(ctx context.Context, pair currency.Pair, assetType asset.Item) (*orderbook.Base, error) {
+	err := ku.CurrencyPairs.IsAssetEnabled(assetType)
+	if err != nil {
+		return nil, err
+	}
+	pair, err = ku.FormatExchangeCurrency(pair, assetType)
+	if err != nil {
+		return nil, err
+	}
+	var ordBook *Orderbook
+	switch assetType {
+	case asset.Futures:
+		ordBook, err = ku.GetFuturesOrderbook(ctx, pair.String())
+	case asset.Spot, asset.Margin:
+		if ku.IsRESTAuthenticationSupported() {
+			ordBook, err = ku.GetOrderbook(ctx, pair.String())
+			if err == nil {
+				break
+			}
```

Comment:
return error

Reply:
It tries to check if the Authenticated endpoint is reachable and uses the orderbook data from private if possible. If not, use the public one.

So, I don't see the relevance of having an error return. 

If needed, I can remove the private GetOrderbook endpoint call.

Reply:
I understand where you are coming from but you are still obfuscating an error. So maybe call this only if keys are set? Then send the request, because if keys are set this shouldn't error, it's not API configurable.